src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -336,12 +337,19 @@ private void WatchForFileSystemEventsThreadStart(object arg)
 
                 for (long i = 0; i < numEvents.ToInt32(); i++)
                 {
-                    Debug.Assert(eventPaths[i].Length > 0, "Empty events are not supported");
-                    Debug.Assert(eventPaths[i][eventPaths[i].Length - 1] != '/', "Trailing slashes on events is not supported");
+                    Debug.Assert(eventPaths[i] != null);
+                    int endPoint = new ReadOnlySpan<byte>(eventPaths[i], 255).IndexOf<byte>(0);
+
+                    Debug.Assert(endPoint > 0, "Empty events are not supported");
+                    ReadOnlySpan<byte> eventBytes = new ReadOnlySpan<byte>(eventPaths[i], endPoint);
+
+                    Span<char> buffer = new char[Encoding.UTF8.GetMaxCharCount(255)];
```

Comment:
Again, this is missing the point. You're allocating on the managed heap. The only "new" you should have in here is for structs- no arrays, no classes.

Reply:
```
Span<char> buffer = stackalloc char[256];
int charCount = Encoding.UTF8.GetChars(eventBytes, buffer);
ReadOnlySpan<char> path = buffer.Slice(0, charCount);
```
is using the stackspace okay ?

Reply:
> is using the stackspace okay ?

No. You could potentially have an 8K buffer here (4K Unicode chars). I would just get your buffers from the ArrayPool to start and investigate adding conditionals to put smaller ones directly on the stack after you get things working. That said, it doesn't even look like you can use stack space as you need to randomly access the data (and you don't want to do multiple conversions, which I've already called out).

Reply:
```C#
private unsafe static void ProcessEvents(char[][] events, byte ** eventPaths, int[] charCounts)
{
    for(int i = 0; i < charCounts.Length; i++ )
    {        
        int endPoint = GetIndex(eventPaths[i]);
        Debug.Assert(endPoint > 0, "Empty events are not supported");        
        events[i] = ArrayPool<char>.Shared.Rent(endPoint);

        // Converting an array of bytes to UTF-8 char array
        fixed(char* temp = &events[i][0])
        {
              charCounts[i] = Encoding.UTF8.GetChars(eventPaths[i], endPoint, temp, endPoint);           
        }
    }
}

private unsafe static int GetIndex(byte* eventPath)
{
    Debug.Assert(eventPath != null);
    int i = 0;

    // Finds the position of null character.    
    while(*eventPath != 0)
    {        
        eventPath++;
        i++;        
    }
    return i;
}

private unsafe void FileSystemEventCallback(
    FSEventStreamRef streamRef, 
    IntPtr clientCallBackInfo, 
    size_t numEvents, 
    byte** eventPaths,
    [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
    Interop.EventStream.FSEventStreamEventFlags[] eventFlags,
    [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
    FSEventStreamEventId[] eventIds)
{
    Debug.Assert((eventPaths != null) && (numEvents.ToInt32() == eventFlags.Length) && (numEvents.ToInt32() == eventIds.Length));
    
    // ... 

    int[] charCounts = new int[numEvents.ToInt32()];
    char[][] events = ArrayPool<char[]>.Shared.Rent(numEvents.ToInt32());
    ProcessEvents(events, eventPaths, charCounts);

    for (long i = 0; i < numEvents.ToInt32(); i++)
    {
        ReadOnlySpan<char> path = new ReadOnlySpan<char>(events[i],0, charCounts[i]);
        Debug.Assert(path[path.Length - 1] != '/', "Trailing slashes on events is not supported");

        // ...
        ReadOnlySpan<char> newPathRelativeName = new ReadOnlySpan<char>(events[pairedId], 0, charCounts[pairedId]).Slice(_fullDirectory.Length);
    }

    for (long i = 0; i < numEvents.ToInt32(); i++)
    {
        ArrayPool<char>.Shared.Return(events[i]);
    }

    ArrayPool<char[]>.Shared.Return(events);
}

```

is this use of arraypool okay ? no 255 limit, converting only once, using arraypool for everything

Reply:
Creating an ArrayPool<char[]> isn't a great idea as it isn't a common type. Putting the list of arrays on the heap isn't a big deal compared to the complexity you'd need to introduce to make this completely stack based- certainly as a first step. The key thing here is not allocating long char arrays that we can't reuse.

I'd also look at putting the rented arrays in Memory<char>. Then you can slice instead of keeping the length separate. You can unwrap the original array to return it using MemoryMarshal.

Reply:
> I'd also look at putting the rented arrays in Memory<char>. Then you can slice instead of keeping the length separate. You can unwrap the original array to return it using MemoryMarshal.

can you please explain this in little bit more detail ?